### ISSUE TITLE
HTTP version error when publishing message via proxy with PublisherClient

### ISSUE THREAD

**Issue by Fedorov1977** [googleapis/google-cloud-dotnet#9801]
We publish messages to PubSub using the PublisherClient. When running from the production environment we set a proxy. However, when the code executes it throws the following error:

**Only HTTP/1.0 and HTTP/1.1 version requests are currently supported. Parameter name: value**

**_Full stack trace:_**
Error: One or more errors occurred. Status(StatusCode="Internal", Detail="Error starting gRPC call. ArgumentException: Only HTTP/1.0 and HTTP/1.1 version requests are currently supported.
Parameter name: value", DebugException="System.ArgumentException: Only HTTP/1.0 and HTTP/1.1 version requests are currently supported.
Parameter name: value
   at System.Net.HttpWebRequest.set_ProtocolVersion(Version value)
   at System.Net.Http.HttpClientHandler.CreateAndPrepareWebRequest(HttpRequestMessage request)
   at System.Net.Http.HttpClientHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Grpc.Net.Client.Internal.GrpcCall`2.<RunCall>d__73.MoveNext() in /_/src/Grpc.Net.Client/Internal/GrpcCall.cs:line 493")

How do I set for example http 1.1 when I create the PublisherClient? The line that throws the exception is:

_return await publisherClient.PublishAsync(message);_

The code currently looks like this (with starting point method PublishMessage:

```
public async Task<string> PublishMessage(string projectId, string topicId, string message)
{
    TopicName topicName = TopicName.FromProjectTopic(projectId, topicId);
    PublisherClient publisherClient = await GetPublisherClient(topicName);       
    
    return await publisherClient.PublishAsync(message);
 }
```

```
 private PublisherClient publisherClient1 = null;
 private async Task<PublisherClient> GetPublisherClient(TopicName topicName)
 {
    if (publisherClient1 == null)
    {
        publisherClient1 = _useProxy ? GetClientWithProxy(topicName) : await PublisherClient.CreateAsync(topicName);
    }

    return publisherClient1;
 }
```

    private PublisherClient GetClientWithProxy(TopicName topicName)
    {
        WebProxy proxy = new WebProxy(_nordnetProxyUrl, _nordnetWebProxyPort);
        PublisherClient publisher = new PublisherClientBuilder
        {
            TopicName = topicName,
            GrpcAdapter = GrpcNetClientAdapter.Default.WithAdditionalOptions(options => options.HttpHandler = new HttpClientHandler
            {
                Proxy = proxy,
                UseProxy = true
            })
        }.Build();

        return publisher;
    }

Some technical information:
* The code is developed using .NET Framework v4.8
* Running on server with OS "Windows Server 2016 Datacenter (have also tried on Windows Server 2019 v1809 with same error message)

I'm using the following versions of Google-libraries in my code (listing all of them 'cause I don't know exactly which ones are relevant):
* Google.Cloud.PubSubV1 - 3.3.0
* Google.Api.CommonProtos - 2.7.0
* Google.Api.Gax - 4.3.0
* Google.Api.Gax.Grcp - 4.3.0
* Google.Apis - 1.58.0
* Google.Apis.Auth - 1.58.0
* Google.Apis.Core - 1.58.0
* Google.Cloud.Iam.V1 - 3.0.0
* Grcp.Auth - 2.51.0
* Grcp.Core - 2.46.5
* Grcp.Core.Api - 2.51.0
* Grcp.Net.Client - 2.51.0
* Grcp.Net.Common - 2.51.0

**Comment by jskeet** [comment]
Please could you post the full stack trace? It's very unclear to me at the moment where that error is coming from. (Could it actually be that your proxy just doesn't support HTTP/2.0?)

Currently our Pub/Sub client library requires gRPC over HTTP/2.0. While it's *possible* that the REST transport may work with Pub/Sub, we haven't enabled or tested that yet. (The Pub/Sub library is complex, and may well rely on aspects of gRPC which aren't perfectly mimicked by the REST transport.)

**Comment by Fedorov1977** [comment]
@jskeet 
Hello, thanks. Yes, I've now updated my post with the stack trace.

**Comment by jskeet** [comment]
> Might that indicate that the proxy is supporting HTTP/2.0?

No, Google.Cloud.Storage.V1 only uses HTTP/1.1 - the Storage API doesn't yet support gRPC.

I suggest you find out which precise proxy you're using, and consult their documentation to find out if they support HTTP/2.0.

It does look like HttpWebRequest.ProtocolVersion only supports 1.0 and 1.1; I don't know enough about the internal details of .NET support for proxies to know how they choose whether to use HttpWebRequest or HttpRequest.

Please could you give details about the platform you're running on and the library versions you're using?

**Comment by Fedorov1977** [comment]
@jskeet 
I've contacted the techinically responsible for the proxy to see which versions of HTTP is supported. Hopefully I will get an answer soon. I've also updated my answer with some technical info, just let me know if you need some more info. Thanks!

**Comment by jskeet** [comment]
It's *possible* that this is a limitation of Grpc.Net.Client on .NET Framework. (You might want to ask on https://github.com/grpc/grpc-dotnet to be sure.) One option would be to use GrpcCoreAdapter instead of GrpcNetClientAdapter, and just use [environment variables](https://github.com/grpc/grpc/blob/master/doc/environment_variables.md) to set the proxy.

**Comment by Fedorov1977** [comment]
Thanks, I'll see if I can figure out how to use the environment variables to set proxy on GrcpCoreAdapter.

What I've also tried is use the publisherClient by not setting a proxy (PublisherClient.CreateAsync(topicName)) and instead configuring the server to use proxy (via proxy settings). I verify that proxy is set by opening a browser and surfing to some site. 

Then, when I execute my program to publish messages, I get the following error (as if the server is unreachable):

Status(StatusCode="DeadlineExceeded", Detail="Deadline Exceeded", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1675161143.746000000","description":"Deadline Exceeded","file":"..\..\..\src\core\ext\filters\deadline\deadline_filter.cc","file_line":81,"grpc_status":4}")

**Comment by jskeet** [comment]
Well when accessing via a browser, I suspect that's at least falling back to using HTTP/1.1. So I think we definitely need to check whether your proxy supports HTTP/2.0 at all.

Are you able to execute simple console applications within the same environment behind a proxy? If so, I can write a simpler app than one using Pub/Sub - while our Pub/Sub library is relatively complex with handwritten parts, we should be able to use another API (e.g. Google.Cloud.Translate.V3) to make requests, and can tinker with that until it works - then apply the same approach to Pub/Sub.

**Comment by Fedorov1977** [comment]
Yes, I can run console apps from the server in question. The application I'm working on for publishin messages is actually a console app that we want to schedule in Windows Task Scheduler.

Much appreciated!

**Comment by jskeet** [comment]
Great, thanks.

Here's a simple console app - just edit the projectId to refer to yours:

```csharp
using Google.Api.Gax.ResourceNames;
using Google.Cloud.Translate.V3;
using System;

// Initially equivalent to LanguageServiceClientBuilder.Create(),
// but we may well want to try different settings.
var client = new TranslationServiceClientBuilder()
{
}.Build();

string projectId = "edit-me";
string locationId = "global";
var request = new TranslateTextRequest
{
    Contents = { "It is raining.", "It is sunny." },
    TargetLanguageCode = "fr-FR",
    ParentAsLocationName = new LocationName(projectId, locationId)
};

var response = client.TranslateText(request);

Console.WriteLine($"Translations returned: {response.Translations.Count}");
Console.WriteLine();
foreach (var translation in response.Translations)
{
    Console.WriteLine($"Detected language: {translation.DetectedLanguageCode}");
    Console.WriteLine($"Translated text: {translation.TranslatedText}");
    Console.WriteLine();
}
```

And the project file (using .NET 4.8):

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net48</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Google.Cloud.Translate.V3" Version="3.1.0" />
  </ItemGroup>

</Project>
```

**Comment by Fedorov1977** [comment]
Awesome! A question has popped up, I ran the code with my credentials (for the production env) from my development machine just to try it out first. My development machine is not blocked by any proxy or anything, I just wanted to test the code from an "unrestricted" environment so to speak.

I got this message:

Status(StatusCode="PermissionDenied", Detail="Cloud Translation API has not been used in project 2140xxxxxx before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/translate.googleapis.com/overview?project=2140xxxxx then retry...[..]

So I opened the suggested url and it says it costs 45 USD/Hour to have it enabled and I don't think I have that possibility.. Or is there any way around it?

**Comment by jskeet** [comment]
Just enabling the API really shouldn't have any cost. Could you let me know where you saw that?

But I'd recommend using a test GCP project anyway, and using credentials for that project in the console app - just to make absolutely certain that you don't accidentally affect production data while testing the connectivity.

**Comment by jskeet** [comment]
Ah - I assume it was in the "enable APIs" section, like this:

<img width="432" alt="image" src="https://user-images.githubusercontent.com/17011/215797072-1a2bf3d0-78e7-45a9-9502-317fa6adab1b.png">

That's the cost of performing AutoML Translation Model Training - which isn't something you'd be doing. The [translation pricing page](https://cloud.google.com/translate/pricing) is clearer.

If you would prefer to use Pub/Sub though, we could use `PublisherApiServiceClient` (which doesn't include hand-written elements) and a "get" operation for a non-existent topic, just to prove connectivity.

**Comment by Fedorov1977** [comment]
@jskeet 

First off: I got response from technicians that the proxy supports both HTTP/1.1 and 2.0..

Now:
I got the translation-app working when connecting to a colleagues personal "test-cloud" and running the app from my unrestricted development machine. 
When moving the app to the server with proxy enabled in windows settings it still couldn't connect. Maybe it couldn't obtain proxy information from the system...?

This is the complete stack trace:
Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="failed to connect to all addresses", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1675254932.617000000","description":"Failed to pick subchannel","file":"..\..\..\src\core\ext\filters\client_channel\client_channel.cc","file_line":3218,"referenced_errors":[{"created":"@1675254932.617000000","description":"failed to connect to all addresses","file":"..\..\..\src\core\lib\transport\error_utils.cc","file_line":165,"grpc_status":14}]}")
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Grpc.Core.Internal.AsyncCall`2.UnaryCall(TRequest msg) in /var/local/git/grpc/src/csharp/Grpc.Core/Internal/AsyncCall.cs:line 78
   at Grpc.Core.Calls.BlockingUnaryCall[TRequest,TResponse](CallInvocationDetails`2 call, TRequest req) in /var/local/git/grpc/src/csharp/Grpc.Core/Calls.cs:line 46
   at Grpc.Core.DefaultCallInvoker.BlockingUnaryCall[TRequest,TResponse](Method`2 method, String host, CallOptions options, TRequest request) in /var/local/git/grpc/src/csharp/Grpc.Core/DefaultCallInvoker.cs:line 46
   at Grpc.Core.Interceptors.InterceptingCallInvoker.<BlockingUnaryCall>b__3_0[TRequest,TResponse](TRequest req, ClientInterceptorContext`2 ctx) in /_/src/Grpc.Core.Api/Interceptors/InterceptingCallInvoker.cs:line 51
   at Grpc.Core.ClientBase.ClientBaseConfiguration.ClientBaseConfigurationInterceptor.BlockingUnaryCall[TRequest,TResponse](TRequest request, ClientInterceptorContext`2 context, BlockingUnaryCallContinuation`2 continuation) in /_/src/Grpc.Core.Api/ClientBase.cs:line 174
   at Grpc.Core.Interceptors.InterceptingCallInvoker.BlockingUnaryCall[TRequest,TResponse](Method`2 method, String host, CallOptions options, TRequest request) in /_/src/Grpc.Core.Api/Interceptors/InterceptingCallInvoker.cs:line 48
   at Google.Cloud.Translate.V3.TranslationService.TranslationServiceClient.TranslateText(TranslateTextRequest request, CallOptions options) in /_/apis/Google.Cloud.Translate.V3/Google.Cloud.Translate.V3/TranslationServiceGrpc.g.cs:line 379
   at Google.Api.Gax.Grpc.ApiCall.GrpcCallAdapter`2.CallSync(TRequest request, CallSettings callSettings) in /_/Google.Api.Gax.Grpc/ApiCall.cs:line 92
   at Google.Api.Gax.Grpc.ApiCallRetryExtensions.<>c__DisplayClass1_0`2.<WithRetry>b__0(TRequest request, CallSettings callSettings) in /_/Google.Api.Gax.Grpc/ApiCallRetryExtensions.cs:line 59
   at Google.Api.Gax.Grpc.ApiCall`2.<>c__DisplayClass12_0.<WithCallSettingsOverlay>b__1(TRequest req, CallSettings cs) in /_/Google.Api.Gax.Grpc/ApiCall.cs:line 187
   at Google.Api.Gax.Grpc.ApiCall`2.Sync(TRequest request, CallSettings perCallCallSettings) in /_/Google.Api.Gax.Grpc/ApiCall.cs:line 150
   at Google.Cloud.Translate.V3.TranslationServiceClientImpl.TranslateText(TranslateTextRequest request, CallSettings callSettings) in /_/apis/Google.Cloud.Translate.V3/Google.Cloud.Translate.V3/TranslationServiceClient.g.cs:line 2891
   at AppGCTranslateTestApp.Program.Execute() in C:\GITTobias\GITTobias\Nordnet\MessageIntegrationService2\GCTranslateTestApp\Program.cs:line 45

**Comment by jskeet** [comment]
@Fedorov1977: Glad to hear we're making progress on the diagnostic side - and that hopefully it won't turn out to be a problem with the proxy itself.

When you ran the failing translation test, had you set any environment variables to specify the proxy, e.g. grpc_proxy, https_proxy, http_proxy?

**Comment by Fedorov1977** [comment]
Hello!
Nothing was set code-wise nor had I specified any system-variables on the server manually. 

After I set the proxy via the proxy-settings-menu on the server I double checked the system variables to see if it creates any variables when the proxy is activated but none of the names you mention exist as system variables.

 (We have a secondary backup-server for load-balancing and this one I can test more freely on because no traffic comes to this server normally. It was when I activated the proxy on the primary server that I noticed that our webapps didn't work and the users were a bit stressed for a few minutes, hehe)

**Comment by jskeet** [comment]
> After I set the proxy via the proxy-settings-menu on the server I double checked the system variables to see if it creates any variables when the proxy is activated but none of the names you mention exist as system variables.

No, I'd expect you to have to set that manually. (You could either do so within Windows, or in the app as the very first line, before gRPC gets initialized.)

**Comment by Fedorov1977** [comment]
Aha, so then my next step would be to set these variables in the system (grpc_proxy, https_proxy, http_proxy) and try executing the app again? (Just so I understand correctly) :)

**Comment by jskeet** [comment]
Yes - but I'd suggest just setting grpc_proxy.

**Comment by Fedorov1977** [comment]
Hello, 

I've tried setting the system variable on the server as per the attached image. It gave the same error message as before (I pasted it below just in case...)

![image](https://user-images.githubusercontent.com/48356015/216332790-adf2f556-a120-4e09-85fc-a6f9cfea9be8.png)

I also tried setting the variable programmatically just like I set the "GOOGLE_APPLICATION_CREDENTIALS":
Environment.SetEnvironmentVariable("grpc_proxy", "extproxy-vip.prod.xxxxxxx.se:3128");
It also resulted in the same error message.

I tried both options both with the proxy turned on and off via Windows Proxy settings just to be sure...

Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="failed to connect to all addresses", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1675343397.787000000","description":"Failed to pick subchannel","file":"..\..\..\src\core\ext\filters\client_channel\client_channel.cc","file_line":3218,"referenced_errors":[{"created":"@1675343397.787000000","description":"failed to connect to all addresses","file":"..\..\..\src\core\lib\transport\error_utils.cc","file_line":165,"grpc_status":14}]}")
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Grpc.Core.Internal.AsyncCall`2.UnaryCall(TRequest msg) in /var/local/git/grpc/src/csharp/Grpc.Core/Internal/AsyncCall.cs:line 78
   at Grpc.Core.Calls.BlockingUnaryCall[TRequest,TResponse](CallInvocationDetails`2 call, TRequest req) in /var/local/git/grpc/src/csharp/Grpc.Core/Calls.cs:line 46
   at Grpc.Core.DefaultCallInvoker.BlockingUnaryCall[TRequest,TResponse](Method`2 method, String host, CallOptions options, TRequest request) in /var/local/git/grpc/src/csharp/Grpc.Core/DefaultCallInvoker.cs:line 46
   at Grpc.Core.Interceptors.InterceptingCallInvoker.<BlockingUnaryCall>b__3_0[TRequest,TResponse](TRequest req, ClientInterceptorContext`2 ctx) in /_/src/Grpc.Core.Api/Interceptors/InterceptingCallInvoker.cs:line 51
   at Grpc.Core.ClientBase.ClientBaseConfiguration.ClientBaseConfigurationInterceptor.BlockingUnaryCall[TRequest,TResponse](TRequest request, ClientInterceptorContext`2 context, BlockingUnaryCallContinuation`2 continuation) in /_/src/Grpc.Core.Api/ClientBase.cs:line 174
   at Grpc.Core.Interceptors.InterceptingCallInvoker.BlockingUnaryCall[TRequest,TResponse](Method`2 method, String host, CallOptions options, TRequest request) in /_/src/Grpc.Core.Api/Interceptors/InterceptingCallInvoker.cs:line 48
   at Google.Cloud.Translate.V3.TranslationService.TranslationServiceClient.TranslateText(TranslateTextRequest request, CallOptions options) in /_/apis/Google.Cloud.Translate.V3/Google.Cloud.Translate.V3/TranslationServiceGrpc.g.cs:line 379
   at Google.Api.Gax.Grpc.ApiCall.GrpcCallAdapter`2.CallSync(TRequest request, CallSettings callSettings) in /_/Google.Api.Gax.Grpc/ApiCall.cs:line 92
   at Google.Api.Gax.Grpc.ApiCallRetryExtensions.<>c__DisplayClass1_0`2.<WithRetry>b__0(TRequest request, CallSettings callSettings) in /_/Google.Api.Gax.Grpc/ApiCallRetryExtensions.cs:line 59
   at Google.Api.Gax.Grpc.ApiCall`2.<>c__DisplayClass12_0.<WithCallSettingsOverlay>b__1(TRequest req, CallSettings cs) in /_/Google.Api.Gax.Grpc/ApiCall.cs:line 187
   at Google.Api.Gax.Grpc.ApiCall`2.Sync(TRequest request, CallSettings perCallCallSettings) in /_/Google.Api.Gax.Grpc/ApiCall.cs:line 150
   at Google.Cloud.Translate.V3.TranslationServiceClientImpl.TranslateText(TranslateTextRequest request, CallSettings callSettings) in /_/apis/Google.Cloud.Translate.V3/Google.Cloud.Translate.V3/TranslationServiceClient.g.cs:line 2891
   at AppGCTranslateTestApp.Program.Execute() in C:\GITTobias\GITTobias\Nordnet\MessageIntegrationService2\GCTranslateTestApp\Program.cs:line

**Comment by jskeet** [comment]
Hmm. Okay, that's good information, even if it's dispiriting. Let's just check that it's using Grpc.Core. I suspect it is, but let's make sure. The simplest way to do that is to set the `GRPC_DEFAULT_ADAPTER_OVERRIDE` environment variable ([docs](https://cloud.google.com/dotnet/docs/reference/help/transports#default-selection)).

If you set it to the value `Grpc.Core` I suspect you'll get this error. If you set it to `Grpc.Net.Client` I suspect you'll get the error we started with. If you could just check both of those to confirm, that would be helpful.

**Comment by Fedorov1977** [comment]
Hello,

With variable set to "Grcp.Core" i get this error:
**Unknown value for environment variable 'GRPC_DEFAULT_ADAPTER_OVERRIDE': 'Grcp.Core'**

With the value set to Grcp.Net.Client I get this error:
**gRPC requires extra configuration on .NET implementations that don't support gRPC over HTTP/2. An HTTP provider must be specified using GrpcChannelOptions.HttpHandler.The configured HTTP provider must either support HTTP/2 or be configured to use gRPC-Web. See https://aka.ms/aspnet/grpc/netstandard for details.**

**Comment by jskeet** [comment]
You've got a typo for the first value: Grcp.Core should be Grpc.Core

Good to know that it's *not* using Grpc.Net.Client though. (We can explore potentially trying to do so later on, but that's a different matter.)

I'm now starting to wonder whether the initial exception was from Auth (triggered by gRPC) rather than from gRPC itself.

Please could you add this code to the example (as a first couple of lines) and let us know the output:

```csharp
using Google.Apis.Auth.OAuth2;

var credential = GoogleCredential.GetApplicationDefault();
Console.WriteLine(credential.UnderlyingCredential.GetType());
```

**Comment by Fedorov1977** [comment]
I'm so sorry, my brain was tired and I tested my "publish messages app" instead of the text translation app.
I have updated my previous answer with results from each app with each setting for the system variable, sorry about that.

**Comment by jskeet** [comment]
You still have a typo with the updated results - you specified Grcp.Net.Client instead of Grpc.Net.Client the second time.

In your original code, you explicitly specified the HttpHandler - but not in the way that https://learn.microsoft.com/en-us/aspnet/core/grpc/netstandard suggests for .NET Framework. Have you tried specifying `WinHttpHandler`?

I'm not sure whether we should focus on trying to get Grpc.Net.Client working as per your original code, or Grpc.Core working (which will remove one area of uncertainty). Do you have reasons to particularly want to use Grpc.Net.Client?

**Comment by Fedorov1977** [comment]
I double checked now and the typo was only in my comment, not the actual variable value fortunately.
The variable value was "Grpc.Net.Client". I re-ran it now just to triple check and the error is "gRPC requires extra configuration..."

I guess you are referring to this code:

```
PublisherClient publisher = new PublisherClientBuilder
    {
        TopicName = topicName,
        GrpcAdapter = GrpcNetClientAdapter.Default.WithAdditionalOptions(options => options.HttpHandler = new HttpClientHandler
        {
            Proxy = proxy,
            UseProxy = true
        })
    }.Build();
```
I don't have any particular reasons to use the Grpc.Net.Client. I can use whatever works so to speak.  :)

**Comment by Fedorov1977** [comment]
Yes, you are right of course. I ran the text-app now with variable value "Grpc.Net.Client" and the error is the "gRPC requires extra configuration.." message. 

The output is:
**Google.Apis.Auth.OAuth2.ServiceAccountCredential**

**Comment by jskeet** [comment]
Okay, so just to be clear, let's:

- Clear the `GRPC_DEFAULT_ADAPTER_OVERRIDE` environment variable or set it to Grpc.Core. It sounds like it shouldn't matter which
- Make sure you've got `grpc_proxy` set appropriately
- Set up additional Grpc.Core logging as per [this documentation](https://cloud.google.com/dotnet/docs/reference/help/troubleshooting#transport-logging-with-grpccore):
  - Add the `GrpcEnvironment.SetLogger(new ConsoleLogger());` call at the start of the code (with required `using` directives)
  - Set the `GRPC_VERBOSITY` environment variable to `debug`
  - Set the `GRPC_TRACE` environment variable to `all`

Then capture the output, which may be large. (I don't know how much Grpc.Core logs if it can't connect.)

If you'd prefer not to put the log in this issue (as it may contain sensitive information which may be hard to spot amongst a lot of logging) please feel free to share it with me (jonskeet@google.com) another way, e.g. via Google Drive.

**Comment by Fedorov1977** [comment]
I've now set the environment variables accordingly:
![image](https://user-images.githubusercontent.com/48356015/216962246-30a15427-a415-47d1-9276-f9d1928f5bbb.png)

There was a lot logged to the output but the application never stopped, or at least it ran for about 5 minutes and then I stopped it manually and restarted it and let it run for about 30 sec.
Maybe it didn't stop because I didn't really get what you meant by "with required using directives".
So this is the code that was actually run, let me know if I should do any changes: (I emailed you the text file from my gmail)

```
static void Main(string[] args)
        {            
            Execute();
        }

        static private void Execute()
        {
            Environment.SetEnvironmentVariable("GOOGLE_APPLICATION_CREDENTIALS", _credentialsSettingsPath);
            GrpcEnvironment.SetLogger(new ConsoleLogger());
            var credential = GoogleCredential.GetApplicationDefault();
            Console.WriteLine(credential.UnderlyingCredential.GetType());

            string variableName = "grpc_proxy";
            try
            {                
                var client = new TranslationServiceClientBuilder()
                {   }.Build();

                string projectId = "primal-fulcrum-360xxx";
                string locationId = "global";

                var request = new TranslateTextRequest
                {
                    Contents = { "It is raining.", "It is sunny." },
                    TargetLanguageCode = "es-ES",
                    ParentAsLocationName = new LocationName(projectId, locationId)
                };

                var response = client.TranslateText(request);

                Console.WriteLine($"Translations returned: {response.Translations.Count}");
                Console.WriteLine();
                foreach (var translation in response.Translations)
                {
                    Console.WriteLine($"Detected language: {translation.DetectedLanguageCode}");
                    Console.WriteLine($"Translated text: {translation.TranslatedText}");
                    Console.WriteLine();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine($"{e}");                                
            }
            Console.ReadKey();
        }
```

**Comment by jskeet** [comment]
By "with required using directives" I mean the lines in the documentation I linked to before:

```csharp
using Grpc.Core;
using Grpc.Core.Logging;
```

I'm not sure why you've added extra methods etc - I'd expected you'd just run `dotnet run` from the command line with exactly the code I provided.

Looking at the log you've sent me, this looks interesting:

```text
..\..\..\src\core\ext\filters\client_channel\http_proxy.cc:79: cannot parse value of 'http_proxy' env var. Error: OK
```

I'm not sure whether that means it's not *using* a proxy. Please could you check the value of that environment variable on the command line?

I'll ask a colleague on the gRPC team to have a look at this issue as well.

**Comment by Fedorov1977** [comment]
That's interesting, when using the StorageClient and setting a proxy for it I didn't set "http://" in the address.
The StorageClient is working fine.

What I tried now was to remove the SetLogger from the code, remove the system variables regarding trace and verbocity and adding "http://" to the proxy address as per the image below:
![image](https://user-images.githubusercontent.com/48356015/216973682-50067b46-3a16-47b9-a9c7-908948998371.png)

The result was that the translated text was returned and logged properly to the console window.

So, obviously I wanted to try my message publishing app now and it actually worked correctly now!
I assume then that the PublisherClient reads this value from the registry and unlike the StorageClient it does not "assume" http:// if it is not part of the given uri for the proxy...
Could that be it?

Anyways, at this moment in time it would seem that my application is working properly thanks to the grpx_proxy environment variable...which is great!

**Comment by jskeet** [comment]
Neither PublisherClient nor StorageClient themselves know anything about the proxies - it's the entirely different technologies of Grpc.Core and HttpClient that are doing registry/environment lookups for the proxy. So it sounds like HttpClient (part of .NET) works with scheme-less proxy specifications, but the C-core part of Grpc.Core doesn't. I'll add a section to our documentation to note this.

**Comment by jskeet** [comment]
Glad everything is resolved though. I'll close this issue now; I should be able to update the troubleshooting docs later in the week.

**Comment by Fedorov1977** [comment]
Sounds good! Thanks so much for the assistance!
In my opinion this matter could be considered "closed". 

Have a great day!
